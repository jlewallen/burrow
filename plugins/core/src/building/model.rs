use crate::library::model::*;

use std::str::FromStr;

#[derive(Clone, Serialize, Deserialize, PartialEq, ToTaggedJson, Reply, Debug)]
#[serde(rename_all = "camelCase")]
pub enum Constructed {
    Area(AreaObservation),
}

impl TryFrom<Constructed> for Effect {
    type Error = TaggedJsonError;

    fn try_from(value: Constructed) -> std::result::Result<Self, Self::Error> {
        Ok(Self::Reply(value.to_tagged_json()?.into()))
    }
}

#[derive(Debug, Serialize, ToTaggedJson)]
#[serde(rename_all = "camelCase")]
struct EditorReply {}

impl Reply for EditorReply {}

#[derive(Default, Clone, Debug)]
pub struct QuickEdit {
    pub name: Option<String>,
    pub desc: Option<String>,
}

impl TryFrom<&EntityPtr> for QuickEdit {
    type Error = DomainError;

    fn try_from(value: &EntityPtr) -> std::result::Result<Self, Self::Error> {
        let name = Some(value.name()?);
        let desc = value.desc()?;
        Ok(Self { name, desc })
    }
}

const SEPARATOR: &str =
    "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~";

impl FromStr for QuickEdit {
    type Err = DomainError;

    fn from_str(s: &str) -> std::result::Result<Self, Self::Err> {
        let parts: Vec<_> = s.split(SEPARATOR).collect();
        if parts.len() != 2 {
            return Err(DomainError::Anyhow(anyhow::anyhow!("malformed quick edit")));
        }

        let (name, desc) = match parts[..] {
            [name, desc] => (name, desc),
            _ => todo!(),
        };

        let name = Some(name.trim().to_owned());
        let desc = Some(desc.trim().to_owned());

        Ok(Self { name, desc })
    }
}

impl ToString for QuickEdit {
    fn to_string(&self) -> String {
        format!(
            "{}\n\n{}\n\n{}",
            self.name.as_deref().unwrap_or(""),
            SEPARATOR,
            self.desc.as_deref().unwrap_or("")
        )
    }
}
