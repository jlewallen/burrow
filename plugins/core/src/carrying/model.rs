use crate::{library::model::*, tools};

pub use kernel::common::Carrying;

#[derive(Debug, Serialize, Deserialize, Default)]
pub struct Containing {
    pub(crate) holding: Vec<EntityRef>,
    pub(crate) capacity: Option<u32>,
    pub(crate) produces: HashMap<String, String>,
}

impl Scope for Containing {
    fn scope_key() -> &'static str {
        "containing"
    }
}

impl Containing {
    pub fn start_carrying(&mut self, item: &EntityPtr) -> Result<bool, DomainError> {
        if let Some(carryable) = item.scope::<Carryable>()? {
            let holding = self
                .holding
                .iter()
                .map(|h| h.to_entity())
                .collect::<Result<Vec<_>, _>>()?;

            for held in holding {
                if is_kind(&held, &carryable.kind)? {
                    let mut combining = held.scope_mut::<Carryable>()?;

                    combining.increase_quantity(&carryable.quantity.into())?;

                    combining.save()?;

                    get_my_session()?.obliterate(item)?;

                    return Ok(true);
                }
            }
        }

        if !self.is_holding(item) {
            self.holding.push(item.entity_ref());
        }

        Ok(true)
    }

    pub fn is_holding(&self, item: &EntityPtr) -> bool {
        self.holding.iter().any(|i| *i.key() == item.key())
    }

    fn remove_item(&mut self, item: &EntityPtr) -> Result<bool, DomainError> {
        self.holding = self
            .holding
            .iter()
            .flat_map(|i| {
                if *i.key() == item.key() {
                    vec![]
                } else {
                    vec![i.clone()]
                }
            })
            .collect::<Vec<EntityRef>>()
            .to_vec();

        Ok(true)
    }

    pub fn stop_carrying(&mut self, found: Found) -> Result<Option<EntityPtr>, DomainError> {
        let item = found.entity()?;

        if !self.is_holding(item) {
            return Ok(None);
        }

        if let Some(carryable) = item.scope::<Carryable>()? {
            match &found {
                Found::One(item) => {
                    if carryable.quantity > 1.0 {
                        let (_original, separated) = tools::separate(item, &1.0.into())?;

                        Ok(Some(separated))
                    } else {
                        self.remove_item(item)?;

                        Ok(Some(item.clone()))
                    }
                }
                Found::Quantified(q, entity) => {
                    if carryable.quantity > q.as_f32() {
                        let (_original, separated) = tools::separate(entity, q)?;

                        Ok(Some(separated))
                    } else {
                        Ok(None)
                    }
                }
            }
        } else {
            self.remove_item(item)?;

            Ok(Some(item.clone()))
        }
    }
}

#[derive(Debug, Serialize, Deserialize)]
pub struct Carryable {
    kind: Kind,
    quantity: f32,
}

fn is_kind(entity: &EntityPtr, kind: &Kind) -> Result<bool, DomainError> {
    if let Some(carryable) = entity.scope::<Carryable>()? {
        Ok(*carryable.kind() == *kind)
    } else {
        Ok(false)
    }
}

impl Default for Carryable {
    fn default() -> Self {
        let session = get_my_session().expect("No session in Entity::new_blank!");
        Self {
            kind: Kind::new(session.new_identity()),
            quantity: 1.0,
        }
    }
}

impl Carryable {
    pub fn quantity(&self) -> f32 {
        self.quantity
    }

    pub fn decrease_quantity(&mut self, q: &Quantity) -> Result<&mut Self, DomainError> {
        self.sanity_check_quantity();

        if *q < 1.0.into() || *q > self.quantity.into() {
            Err(DomainError::Impossible)
        } else {
            self.quantity -= q.as_f32();

            Ok(self)
        }
    }

    pub fn increase_quantity(&mut self, q: &Quantity) -> Result<&mut Self, DomainError> {
        self.sanity_check_quantity();

        self.quantity += q.as_f32();

        Ok(self)
    }

    pub fn set_quantity(&mut self, q: &Quantity) -> Result<&mut Self, DomainError> {
        self.quantity = q.as_f32();

        Ok(self)
    }

    pub fn kind(&self) -> &Kind {
        &self.kind
    }

    pub fn set_kind(&mut self, kind: &Kind) {
        self.kind = kind.clone();
    }

    // Migrate items that were initialized with 0 quantities.
    fn sanity_check_quantity(&mut self) {
        if self.quantity < 1.0 {
            self.quantity = 1.0
        }
    }
}

impl Scope for Carryable {
    fn scope_key() -> &'static str {
        "carryable"
    }
}
