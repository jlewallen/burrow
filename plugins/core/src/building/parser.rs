use crate::library::parser::*;

use super::actions::{
    AddScopeAction, BidirectionalDigAction, BuildAreaAction, DuplicateAction, EditAction,
    EditRawAction, LimboAction, MakeItemAction, ObliterateAction,
};

pub struct EditActionParser {}

impl ParsesActions for EditActionParser {
    fn try_parse_action(&self, i: &str) -> EvaluationResult {
        let (_, action) = alt((
            map(
                preceded(pair(tag("edit raw"), spaces), noun_or_specific),
                |item| -> Box<dyn Action> { Box::new(EditRawAction { item }) },
            ),
            map(
                preceded(pair(tag("edit"), spaces), noun_or_specific),
                |item| -> Box<dyn Action> { Box::new(EditAction { item }) },
            ),
        ))(i)?;

        Ok(Some(action))
    }
}

pub struct MakeItemParser {}

impl ParsesActions for MakeItemParser {
    fn try_parse_action(&self, i: &str) -> EvaluationResult {
        let (_, action) = map(
            preceded(
                tuple((tag("@make"), spaces, tag("item"), spaces)),
                string_literal,
            ),
            |name| MakeItemAction { name: name.into() },
        )(i)?;

        Ok(Some(Box::new(action)))
    }
}

pub struct DuplicateActionParser {}

impl ParsesActions for DuplicateActionParser {
    fn try_parse_action(&self, i: &str) -> EvaluationResult {
        let (_, action) = map(
            preceded(pair(tag("@duplicate"), spaces), noun_or_specific),
            |item| DuplicateAction { item },
        )(i)?;

        Ok(Some(Box::new(action)))
    }
}

pub struct LimboActionParser {}

impl ParsesActions for LimboActionParser {
    fn try_parse_action(&self, i: &str) -> EvaluationResult {
        let (_, action) = map(tag("@limbo"), |_| LimboAction {})(i)?;

        Ok(Some(Box::new(action)))
    }
}

pub struct ObliterateActionParser {}

impl ParsesActions for ObliterateActionParser {
    fn try_parse_action(&self, i: &str) -> EvaluationResult {
        let (_, action) = map(
            preceded(pair(tag("@obliterate"), spaces), noun_or_specific),
            |item| ObliterateAction { item },
        )(i)?;

        Ok(Some(Box::new(action)))
    }
}

pub struct BidirectionalDigActionParser {}

impl ParsesActions for BidirectionalDigActionParser {
    fn try_parse_action(&self, i: &str) -> EvaluationResult {
        let (_, action) = map(
            tuple((
                preceded(pair(tag("@dig"), spaces), string_literal),
                preceded(pair(spaces, pair(tag("to"), spaces)), string_literal),
                preceded(pair(spaces, pair(tag("for"), spaces)), string_literal),
            )),
            |(outgoing, returning, new_area)| BidirectionalDigAction {
                outgoing: outgoing.into(),
                returning: returning.into(),
                new_area: new_area.into(),
            },
        )(i)?;

        Ok(Some(Box::new(action)))
    }
}

pub struct ScopeActionParser {}

impl ParsesActions for ScopeActionParser {
    fn try_parse_action(&self, i: &str) -> EvaluationResult {
        let (_, action) = map(
            preceded(pair(tag("@scope"), spaces), camel_case_word),
            |scope_key| AddScopeAction {
                scope_key: scope_key.to_owned(),
            },
        )(i)?;

        Ok(Some(Box::new(action)))
    }
}

pub struct BuildAreaParser {}

impl ParsesActions for BuildAreaParser {
    fn try_parse_action(&self, i: &str) -> EvaluationResult {
        let (_, action) = map(
            preceded(tuple((tag("@build"), spaces)), string_literal),
            |name| BuildAreaAction { name: name.into() },
        )(i)?;

        Ok(Some(Box::new(action)))
    }
}
