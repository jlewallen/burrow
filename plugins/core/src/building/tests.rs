use super::parser::*;
use super::*;
use crate::building::actions::{SaveEntityJsonAction, SaveQuickEditAction};
use crate::building::model::QuickEdit;
use crate::fashion::model::Wearable;
use crate::library::tests::*;
use crate::{
    {carrying::model::Containing, looking::model::new_area_observation, tools},
    {BuildSurroundings, QuickThing},
};

#[test]
fn it_fails_to_edit_unknown_items() -> Result<()> {
    let mut build = BuildSurroundings::new()?;
    let (session, surroundings) = build
        .ground(vec![QuickThing::Object("Cool Broom")])
        .build()?;

    let action = try_parsing(EditActionParser {}, "edit rake")?;
    let action = action.unwrap();
    let reply = action.perform(session, &surroundings)?;

    let reply: SimpleReply = reply.json_as()?;
    assert_eq!(reply, SimpleReply::NotFound);

    Ok(())
}

#[test]
fn it_fails_to_edit_raw_unknown_items() -> Result<()> {
    let mut build = BuildSurroundings::new()?;
    let (session, surroundings) = build
        .ground(vec![QuickThing::Object("Cool Broom")])
        .build()?;

    let action = try_parsing(EditActionParser {}, "edit raw rake")?;
    let action = action.unwrap();
    let reply = action.perform(session, &surroundings)?;

    let reply: SimpleReply = reply.json_as()?;
    assert_eq!(reply, SimpleReply::NotFound);

    Ok(())
}

#[test]
fn it_edits_items_named() -> Result<()> {
    let mut build = BuildSurroundings::new()?;
    let (session, surroundings) = build
        .ground(vec![QuickThing::Object("Cool Broom")])
        .build()?;

    let action = try_parsing(EditActionParser {}, "edit broom")?;
    let action = action.unwrap();
    let reply = action.perform(session, &surroundings)?;

    insta::assert_json_snapshot!(reply.to_debug_json()?);

    Ok(())
}

#[test]
fn it_edits_raw_items_named() -> Result<()> {
    let mut build = BuildSurroundings::new()?;
    let (session, surroundings) = build
        .ground(vec![QuickThing::Object("Cool Broom")])
        .build()?;

    let action = try_parsing(EditActionParser {}, "edit raw broom")?;
    let action = action.unwrap();
    let reply = action.perform(session, &surroundings)?;

    insta::assert_json_snapshot!(reply.to_debug_json()?);

    Ok(())
}

#[test]
fn it_fails_to_edit_items_by_missing_gid() -> Result<()> {
    let mut build = BuildSurroundings::new()?;
    let (session, surroundings) = build
        .ground(vec![QuickThing::Object("Cool Broom")])
        .build()?;

    let action = try_parsing(EditActionParser {}, "edit #1201")?;
    let action = action.unwrap();
    let reply = action.perform(session, &surroundings)?;

    let reply: SimpleReply = reply.json_as()?;
    assert_eq!(reply, SimpleReply::NotFound);

    Ok(())
}

#[test]
fn it_edits_items_by_gid() -> Result<()> {
    let mut build = BuildSurroundings::new()?;
    let (session, surroundings) = build
        .ground(vec![QuickThing::Object("Cool Broom")])
        .build()?;

    let action = try_parsing(EditActionParser {}, "edit #2")?;
    let action = action.unwrap();
    let reply = action.perform(session, &surroundings)?;

    insta::assert_json_snapshot!(reply.to_debug_json()?);

    Ok(())
}

#[test]
fn it_fails_to_duplicate_unknown_items() -> Result<()> {
    let mut build = BuildSurroundings::new()?;
    let (session, surroundings) = build
        .ground(vec![QuickThing::Object("Cool Broom")])
        .build()?;

    let action = try_parsing(DuplicateActionParser {}, "@duplicate rake")?;
    let action = action.unwrap();
    let reply = action.perform(session, &surroundings)?;

    let reply: SimpleReply = reply.json_as()?;
    assert_eq!(reply, SimpleReply::NotFound);

    Ok(())
}

#[test]
fn it_duplicates_items_named() -> Result<()> {
    let mut build = BuildSurroundings::new()?;
    let (session, surroundings) = build
        .hands(vec![QuickThing::Object("Cool Broom")])
        .build()?;

    let action = try_parsing(DuplicateActionParser {}, "@duplicate broom")?;
    let action = action.unwrap();
    let reply = action.perform(session.clone(), &surroundings)?;
    let (_world, person, _area) = surroundings.unpack();

    let reply: SimpleReply = reply.json_as()?;
    assert_eq!(reply, SimpleReply::Done);
    assert_eq!(person.scope::<Containing>()?.holding.len(), 1);
    assert_eq!(
        tools::quantity(&person.scope::<Containing>()?.holding[0].to_entry()?)?,
        2.0
    );

    Ok(())
}

#[test]
fn it_fails_to_obliterate_unknown_items() -> Result<()> {
    let mut build = BuildSurroundings::new()?;
    let (session, surroundings) = build
        .hands(vec![QuickThing::Object("Cool Broom")])
        .build()?;

    let action = try_parsing(ObliterateActionParser {}, "@obliterate rake")?;
    let action = action.unwrap();
    let reply = action.perform(session, &surroundings)?;

    let reply: SimpleReply = reply.json_as()?;
    assert_eq!(reply, SimpleReply::NotFound);

    Ok(())
}

#[test]
fn it_obliterates_items_named() -> Result<()> {
    let mut build = BuildSurroundings::new()?;
    let (session, surroundings) = build
        .hands(vec![QuickThing::Object("Cool Broom")])
        .build()?;

    let action = try_parsing(ObliterateActionParser {}, "@obliterate broom")?;
    let action = action.unwrap();
    let reply = action.perform(session.clone(), &surroundings)?;
    let (_world, person, area) = surroundings.unpack();

    let reply: SimpleReply = reply.json_as()?;
    assert_eq!(reply, SimpleReply::Done);
    // It's not enough just to check this, but why not given how easy.
    // Should actually verify it's deleted.
    assert_eq!(person.scope::<Containing>()?.holding.len(), 0);
    assert_eq!(area.scope::<Containing>()?.holding.len(), 0);

    build.flush()?;

    Ok(())
}

#[test]
fn it_digs_bidirectionally() -> Result<()> {
    let mut build = BuildSurroundings::new()?;
    let (session, surroundings) = build.plain().build()?;

    let action = try_parsing(
        BidirectionalDigActionParser {},
        r#"@dig "North Exit" to "South Exit" for "New Area""#,
    )?;
    let action = action.unwrap();
    let reply = action.perform(session.clone(), &surroundings)?;
    let (_, living, _area) = surroundings.unpack();

    // Not the best way of finding the constructed area.
    let destination = session
        .entry(&LookupBy::Gid(&EntityGid::new(4)))?
        .ok_or(DomainError::EntityNotFound)?;

    let reply: AreaObservation = reply.json_as()?;
    assert_eq!(reply, new_area_observation(&living, &destination)?);

    Ok(())
}

#[test]
fn it_makes_items() -> Result<()> {
    let mut build = BuildSurroundings::new()?;
    let (session, surroundings) = build.plain().build()?;

    let action = try_parsing(MakeItemParser {}, r#"@make item "Blue Rake""#)?;
    let action = action.unwrap();
    let reply = action.perform(session.clone(), &surroundings)?;
    let (_, living, _area) = surroundings.unpack();

    let reply: SimpleReply = reply.json_as()?;
    assert_eq!(reply, SimpleReply::Done);

    assert_eq!(living.scope::<Containing>()?.holding.len(), 1);

    Ok(())
}

#[test]
fn it_saves_changes_to_description() -> Result<()> {
    let mut build = BuildSurroundings::new()?;
    let (session, surroundings) = build.plain().build()?;
    let (_world, _living, _area) = surroundings.unpack();

    let description = "Would be really weird if this was the original description".to_owned();
    let mut quick_edit = QuickEdit::default();
    quick_edit.name = Some("NAME!".to_owned());
    quick_edit.desc = Some(description.to_owned());

    let action = Box::new(SaveQuickEditAction {
        key: EntityKey::new("world"),
        copy: WorkingCopy::Markdown(quick_edit.to_string()),
    });
    let reply = action.perform(session.clone(), &surroundings)?;
    let (world, _living, _area) = surroundings.unpack();

    let reply: SimpleReply = reply.json_as()?;
    assert_eq!(reply, SimpleReply::Done);

    assert_eq!(world.desc()?.unwrap(), description.as_str());

    Ok(())
}

#[test]
fn it_saves_changes_to_whole_entities() -> Result<()> {
    let mut build = BuildSurroundings::new()?;
    let (session, surroundings) = build.plain().build()?;
    let (_, living, area) = surroundings.unpack();

    let original = living.entity();
    let original = original.to_json_value()?;

    let action = Box::new(SaveEntityJsonAction {
        key: area.key().clone(),
        copy: WorkingCopy::Json(original),
    });
    let reply = action.perform(session.clone(), &surroundings)?;
    let (_, _living, _area) = surroundings.unpack();

    let reply: SimpleReply = reply.json_as()?;
    assert_eq!(reply, SimpleReply::Done);

    // TODO Would be really nice to have some assurances here, even though
    // I'm wondering how often this will actually get used.

    Ok(())
}

#[test]
fn it_adds_scopes_to_solo_held_items() -> Result<()> {
    let mut build = BuildSurroundings::new()?;
    let jacket = build.entity()?.named("Jacket")?.carryable()?.into_entry()?;

    assert!(!jacket.has_scope::<Wearable>()?);

    let (session, surroundings) = build
        .plain()
        .hands(vec![QuickThing::Actual(jacket.clone())])
        .build()?;

    let action = try_parsing(ScopeActionParser {}, r#"@scope wearable"#)?.unwrap();
    let reply = action.perform(session.clone(), &surroundings)?;
    let (_, living, _area) = surroundings.unpack();

    let reply: SimpleReply = reply.json_as()?;
    assert_eq!(reply, SimpleReply::Done);

    assert_eq!(living.scope::<Containing>()?.holding.len(), 1);

    assert!(jacket.has_scope::<Wearable>()?);

    Ok(())
}
