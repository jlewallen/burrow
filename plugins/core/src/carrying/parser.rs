use super::actions::*;
use crate::library::parser::*;

pub struct HoldActionParser {}

impl ParsesActions for HoldActionParser {
    fn try_parse_action(&self, i: &str) -> EvaluationResult {
        let (_, action) = map(
            separated_pair(tag("hold"), spaces, alt((quantified, noun))),
            |(_, item)| HoldAction { item },
        )(i)?;

        Ok(Some(Box::new(action)))
    }
}

pub struct DropActionParser {}

impl ParsesActions for DropActionParser {
    fn try_parse_action(&self, i: &str) -> EvaluationResult {
        let specific = map(
            separated_pair(tag("drop"), spaces, alt((quantified, noun))),
            |(_, item)| DropAction {
                maybe_item: Some(Item::Held(Box::new(item))),
            },
        );

        let everything = map(tag("drop"), |_| DropAction { maybe_item: None });

        let (_, action) = alt((specific, everything))(i)?;

        Ok(Some(Box::new(action)))
    }
}

pub struct TakeOutActionParser {}

impl ParsesActions for TakeOutActionParser {
    fn try_parse_action(&self, i: &str) -> EvaluationResult {
        let item = map(separated_pair(tag("take"), spaces, noun), |(_, item)| item);

        let (_, action) = map(
            separated_pair(separated_pair(item, spaces, tag("out of")), spaces, noun),
            |(item, vessel)| TakeOutAction {
                item: Item::Contained(Box::new(item.0)),
                vessel: vessel,
            },
        )(i)?;

        Ok(Some(Box::new(action)))
    }
}

pub struct PutInsideActionParser {}

impl ParsesActions for PutInsideActionParser {
    fn try_parse_action(&self, i: &str) -> EvaluationResult {
        let item = map(separated_pair(tag("put"), spaces, noun), |(_, item)| item);

        let (_, action) = map(
            separated_pair(
                separated_pair(
                    item,
                    spaces,
                    pair(tag("inside"), opt(pair(spaces, tag("of")))),
                ),
                spaces,
                noun,
            ),
            |(item, vessel)| PutInsideAction {
                item: item.0,
                vessel: vessel,
            },
        )(i)?;

        Ok(Some(Box::new(action)))
    }
}

pub struct GiveToActionParser {}

impl ParsesActions for GiveToActionParser {
    fn try_parse_action(&self, i: &str) -> EvaluationResult {
        let (_, action) = map(
            separated_pair(
                preceded(tag("give"), preceded(spaces, noun)),
                spaces,
                preceded(tag("to"), preceded(spaces, person)),
            ),
            |(item, receiver)| GiveToAction {
                item: Item::Held(Box::new(item)),
                receiver,
            },
        )(i)?;

        Ok(Some(Box::new(action)))
    }
}
