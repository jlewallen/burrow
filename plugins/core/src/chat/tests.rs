use super::parser::*;
use crate::library::tests::*;

#[test]
fn it_raises_conversation_events() -> Result<()> {
    let (_surroundings, effect) = parse_and_perform(SpeakActionParser {}, "say hello, everyone!")?;

    assert!(matches!(effect, Effect::Ok));

    Ok(())
}
