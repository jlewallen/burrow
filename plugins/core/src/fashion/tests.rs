use super::model::*;
use super::parser::*;
use super::*;
use crate::carrying::model::Containing;
use crate::library::tests::*;

#[test]
fn it_wears_unworn_items() -> Result<()> {
    let mut build = BuildSurroundings::new()?;
    let (session, surroundings) = build
        .hands(vec![QuickThing::Wearable("Cool Jacket")])
        .build()?;

    let (_, person, _area) = surroundings.unpack();
    assert_eq!(person.scope::<Wearing>()?.wearing.len(), 0);
    assert_eq!(person.scope::<Containing>()?.holding.len(), 1);

    let action = try_parsing(WearActionParser {}, "wear jacket")?;
    let action = action.unwrap();
    let reply = action.perform(session.clone(), &surroundings)?;

    let reply: SimpleReply = reply.json_as()?;
    assert_eq!(reply, SimpleReply::Done);

    assert_eq!(person.scope::<Wearing>()?.wearing.len(), 1);
    assert_eq!(person.scope::<Containing>()?.holding.len(), 0);

    build.close()?;

    Ok(())
}

#[test]
fn it_removes_worn_items() -> Result<()> {
    let mut build = BuildSurroundings::new()?;
    let (session, surroundings) = build
        .wearing(vec![QuickThing::Wearable("Cool Jacket")])
        .build()?;

    let action = try_parsing(RemoveActionParser {}, "remove jacket")?;
    let action = action.unwrap();
    let reply = action.perform(session.clone(), &surroundings)?;
    let (_, person, _area) = surroundings.unpack();

    let reply: SimpleReply = reply.json_as()?;
    assert_eq!(reply, SimpleReply::Done);

    assert_eq!(person.scope::<Wearing>()?.wearing.len(), 0);
    assert_eq!(person.scope::<Containing>()?.holding.len(), 1);

    build.close()?;

    Ok(())
}
