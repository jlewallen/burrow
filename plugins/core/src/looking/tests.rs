use super::model::*;
use super::parser::LookActionParser;
use super::*;
use crate::library::plugin::try_parsing;
use crate::library::tests::*;

#[test]
fn it_looks_in_empty_area() -> Result<()> {
    let mut build = BuildSurroundings::new()?;
    let (session, surroundings) = build.plain().build()?;

    let action = try_parsing(LookActionParser {}, "look")?;
    let action = action.unwrap();
    let reply = action.perform(session.clone(), &surroundings)?;
    let (_, _person, _area) = surroundings.unpack();

    insta::assert_json_snapshot!(reply.to_debug_json()?);

    build.close()?;

    Ok(())
}

#[test]
fn it_looks_in_area_with_items_on_ground() -> Result<()> {
    let mut build = BuildSurroundings::new()?;
    let (session, surroundings) = build
        .ground(vec![QuickThing::Object("Cool Rake")])
        .ground(vec![QuickThing::Object("Boring Shovel")])
        .build()?;

    let action = try_parsing(LookActionParser {}, "look")?;
    let action = action.unwrap();
    let reply = action.perform(session.clone(), &surroundings)?;
    let (_, _person, _area) = surroundings.unpack();

    insta::assert_json_snapshot!(reply.to_debug_json()?);

    build.close()?;

    Ok(())
}

#[test]
fn it_looks_in_area_with_items_on_ground_and_a_route() -> Result<()> {
    let mut build = BuildSurroundings::new()?;
    let destination = build.make(QuickThing::Place("Place"))?;
    let (session, surroundings) = build
        .ground(vec![QuickThing::Object("Cool Rake")])
        .ground(vec![QuickThing::Object("Boring Shovel")])
        .route("East Exit", QuickThing::Actual(destination))
        .build()?;

    let action = try_parsing(LookActionParser {}, "look")?;
    let action = action.unwrap();
    let reply = action.perform(session.clone(), &surroundings)?;
    let (_, _person, _area) = surroundings.unpack();

    insta::assert_json_snapshot!(reply.to_debug_json()?);

    build.close()?;

    Ok(())
}

#[test]
fn it_looks_in_area_with_items_on_ground_and_holding_items() -> Result<()> {
    let mut build = BuildSurroundings::new()?;
    let destination = build.make(QuickThing::Place("Place"))?;
    let (session, surroundings) = build
        .ground(vec![QuickThing::Object("Boring Shovel")])
        .hands(vec![QuickThing::Object("Cool Rake")])
        .route("East Exit", QuickThing::Actual(destination))
        .build()?;

    let action = try_parsing(LookActionParser {}, "look")?;
    let action = action.unwrap();
    let reply = action.perform(session.clone(), &surroundings)?;
    let (_, _person, _area) = surroundings.unpack();

    insta::assert_json_snapshot!(reply.to_debug_json()?);

    build.close()?;

    Ok(())
}

#[test]
fn it_fails_to_look_inside_non_containers() -> Result<()> {
    let mut build = BuildSurroundings::new()?;
    let (session, surroundings) = build.hands(vec![QuickThing::Object("Not A Box")]).build()?;

    let action = try_parsing(LookActionParser {}, "look inside box")?;
    let action = action.unwrap();
    let reply = action.perform(session.clone(), &surroundings)?;
    let (_, _person, _area) = surroundings.unpack();

    insta::assert_json_snapshot!(reply.to_debug_json()?);

    build.close()?;

    Ok(())
}

#[test]
fn it_looks_inside_containers() -> Result<()> {
    let mut build = BuildSurroundings::new()?;
    let vessel = build
        .entity()?
        .named("Vessel")?
        .carryable()?
        .holding(&vec![build.make(QuickThing::Object("Key"))?])?
        .into_entry()?;
    let (session, surroundings) = build.hands(vec![QuickThing::Actual(vessel)]).build()?;

    let action = try_parsing(LookActionParser {}, "look inside vessel")?;
    let action = action.unwrap();
    let reply = action.perform(session.clone(), &surroundings)?;
    let (_, _person, _area) = surroundings.unpack();

    insta::assert_json_snapshot!(reply.to_debug_json()?);

    build.close()?;

    Ok(())
}

#[test]
fn it_fails_to_look_at_not_found_entities() -> Result<()> {
    let mut build = BuildSurroundings::new()?;
    let vessel = build.entity()?.named("Hammer")?.carryable()?.into_entry()?;
    let (session, surroundings) = build.hands(vec![QuickThing::Actual(vessel)]).build()?;

    let action = try_parsing(LookActionParser {}, "look at shovel")?;
    let action = action.unwrap();
    let reply = action.perform(session.clone(), &surroundings)?;
    let (_, _person, _area) = surroundings.unpack();

    insta::assert_json_snapshot!(reply.to_debug_json()?);

    build.close()?;

    Ok(())
}

#[test]
fn it_looks_at_entities() -> Result<()> {
    let mut build = BuildSurroundings::new()?;
    let hammer = build.entity()?.named("Hammer")?.carryable()?.into_entry()?;
    let (session, surroundings) = build.hands(vec![QuickThing::Actual(hammer)]).build()?;

    let action = try_parsing(LookActionParser {}, "look at hammer")?;
    let action = action.unwrap();
    let reply = action.perform(session.clone(), &surroundings)?;
    let (_, _person, _area) = surroundings.unpack();

    insta::assert_json_snapshot!(reply.to_debug_json()?);

    build.close()?;

    Ok(())
}

#[test]
fn it_sees_worm_items() -> Result<()> {
    let mut build = BuildSurroundings::new()?;
    let jacket = build
        .entity()?
        .named("Jacket")?
        .wearable()?
        .carryable()?
        .into_entry()?;
    let (session, surroundings) = build.wearing(vec![QuickThing::Actual(jacket)]).build()?;

    let action = try_parsing(LookActionParser {}, "look at myself")?;
    let action = action.unwrap();
    let reply = action.perform(session.clone(), &surroundings)?;
    let (_, _person, _area) = surroundings.unpack();

    insta::assert_json_snapshot!(reply.to_debug_json()?);

    build.close()?;

    Ok(())
}

#[test]
fn qualify_name_basics() {
    // Not going to test all of indefinite's behavior here, just build edge
    // cases in our integrating logic.
    assert_eq!(Unqualified::Quantity(1.0, "box").qualify(), "a box");
    assert_eq!(Unqualified::Quantity(2.0, "box").qualify(), "2 boxes");
    assert_eq!(Unqualified::Quantity(1.0, "person").qualify(), "a person");
    assert_eq!(Unqualified::Quantity(2.0, "person").qualify(), "2 people");
    assert_eq!(Unqualified::Quantity(1.0, "orange").qualify(), "an orange");
    assert_eq!(Unqualified::Quantity(2.0, "orange").qualify(), "2 oranges");
    assert_eq!(
        Unqualified::Quantity(1.0, "East Exit").qualify(),
        "an East Exit"
    );
    assert_eq!(Unqualified::Living("Jacob").qualify(), "Jacob");
}
