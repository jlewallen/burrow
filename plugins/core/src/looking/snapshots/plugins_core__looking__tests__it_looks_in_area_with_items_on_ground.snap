---
source: plugins/core/src/looking/tests.rs
expression: reply.to_debug_json()?
---
{
  "areaObservation": {
    "area": {
      "desc": null,
      "key": "E-2",
      "name": "Welcome Area",
      "qualified": "Welcome Area"
    },
    "carrying": [],
    "items": [
      {
        "desc": null,
        "key": "E-3",
        "name": "Cool Rake",
        "qualified": "a Cool Rake"
      },
      {
        "desc": null,
        "key": "E-4",
        "name": "Boring Shovel",
        "qualified": "a Boring Shovel"
      }
    ],
    "living": [
      {
        "desc": null,
        "key": "E-1",
        "name": "Living",
        "qualified": "Living"
      }
    ],
    "person": {
      "desc": null,
      "key": "E-1",
      "name": "Living",
      "qualified": "Living"
    },
    "routes": []
  }
}
